x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/datetime/DateTrunc.java

```diff
@@ -127,9 +234,44 @@ public static Part resolve(String truncateTo) {
             return DateTimeField.resolveMatch(NAME_TO_PART, truncateTo);
         }
 
-        public ZonedDateTime truncate(ZonedDateTime dateTime) {
+        public Object truncate(Object dateTime) {
             return truncateFunction.apply(dateTime);
         }
+
+        private static Object truncateIntervalSmallerThanWeek(Object r, ChronoUnit unit){
+            if (r instanceof IntervalDayTime) {
+                Duration d = ((IntervalDayTime) r).interval();
+                int isNegative = 1;
+                if (d.isNegative()) {
+                    d = d.negated();
+                    isNegative = -1;
+                }
+                long durationInSec = d.getSeconds();
+                long day = durationInSec / SECONDS_PER_DAY;
+                System.out.println(day);
+                durationInSec = durationInSec % SECONDS_PER_DAY;
+                long hour = durationInSec / SECONDS_PER_HOUR;
+                durationInSec = durationInSec % SECONDS_PER_HOUR;
+                long min = durationInSec / SECONDS_PER_MINUTE;
+                durationInSec = durationInSec % SECONDS_PER_MINUTE;
+                long sec = durationInSec;
+                long miliseccond = TimeUnit.NANOSECONDS.toMillis(d.getNano());
+                Duration newDuration = Duration.ZERO;
+                if(unit.ordinal() <= ChronoUnit.DAYS.ordinal())
+                    newDuration = newDuration.plusDays(day*isNegative);
+                if(unit.ordinal() <= ChronoUnit.HOURS.ordinal())
+                    newDuration = newDuration.plusHours(hour*isNegative);
+                if(unit.ordinal() <= ChronoUnit.MINUTES.ordinal())
+                    newDuration = newDuration.plusMinutes(min*isNegative);
+                if(unit.ordinal() <= ChronoUnit.SECONDS.ordinal())
+                    newDuration = newDuration.plusSeconds(sec*isNegative);
+                if(unit.ordinal() <= ChronoUnit.MILLIS.ordinal())
+                    newDuration = newDuration.plusMillis(miliseccond*isNegative);
```

Comment:
Can you explain why did you go with this approach (negating the value, then calculating each time unit, reconstructing the Duration and then negating it again)?
Why not something like the code below?

```
                Duration truncatedDuration = Duration.ZERO;
                switch (unit) {
                    case DAYS: truncatedDuration = Duration.ofDays(d.toDays());
                        break;
                    case HOURS: truncatedDuration = Duration.ofHours(d.toHours());
                        break;
                    case MINUTES: truncatedDuration = Duration.ofMinutes(d.toMinutes());
                        break;
                    case SECONDS: truncatedDuration = Duration.ofSeconds(d.getSeconds());
                        break;
                    case MILLIS: truncatedDuration = Duration.ofMillis(d.toMillis());
                        break;    
                }
```

Reply:
@astefan 
Ofc,
 The return value is of type date/datetime/interval with all fields that are less significant than the selected one set to zero. That's why we need days, hours, minutes, seconds, millis fields of the date/datetime/interval seperatly. For example for 'seconds' we cant use switch case because we need "day", "hour", "minutes" fields of the date/datetime/interval. I think if statements fits these approach. Also, I think we should not use d.toDays(), d.toHours(), etc. functions to get fields.
I want to explain on these example : "DATETRUNC("minutes", INTERVAL 1 days 2 hours 20 minutes 16 seconds")
d.toDays() will return 1 day
d.toHours() will return 26 hours
d.toHours() will return 26*60 second...
But we need integer values 1 for day, 2 for hours, 20 for minutes to truncate.
For negating part, I tried truncate the intervals without negating but it doesnt work for some edge cases. For example, you can construct a duration with "Duration.ofDays(-534).minusHours(23).minusMinutes(59).minusSeconds(59).minusNanos(245000000) " and if you try to getDays with d.toDays() or the way i used without negating, it should 
return -535. Thats why i used these approach. Also i examined the project. You can see, the approach also use in printing intervals in StringUtils class.